tensorflow_model_optimization/python/core/clustering/keras/cluster.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Clustering API functions for Keras models."""
+
+from tensorflow.python import keras
+from tensorflow.python.keras import backend as k
+from tensorflow.python.keras import initializers
+from tensorflow.python.keras.engine.base_layer import Layer
+from tensorflow.python.keras.engine.input_layer import InputLayer
+from tensorflow.python.keras.utils.generic_utils import custom_object_scope
+
+from tensorflow_model_optimization.python.core.clustering.keras import clustering_centroids
+from tensorflow_model_optimization.python.core.clustering.keras.cluster_wrapper import ClusterWeights
+
+
+def clustering_scope():
+  """Provides a scope in which Clustered layers and models can be deserialized.
+
+  If a keras model or layer has been clustered, it needs to be within this scope
+  to be successfully deserialized.
+
+  Returns:
+      Object of type `CustomObjectScope` with clustering objects included.
+
+  Example:
+
+  ```python
+  clustered_model = cluster_weights(model, **self.params)
+  keras.models.save_model(clustered_model, keras_file)
+
+  with clustering_scope():
+    loaded_model = keras.models.load_model(keras_file)
+  ```
+  """
+  return custom_object_scope(
+    {
+      'ClusterWeights': ClusterWeights,
+    }
+  )
+
+
+def cluster_weights(to_cluster, number_of_clusters, cluster_centroids_init, **kwargs):
+  """Modify a keras layer or model to be clustered during training.
+
+  This function wraps a keras model or layer with clustering functionality
+  which clusters the layer's weights during training. For examples, using
+  this with number of clusters equals 8 will ensure that each weight tensor has
+  no more than 8 unique values.
+
+  Before passing to the clustering API, a model should already be trained and show some acceptable performance
```

Comment:
Yes, the concept of clustering (at least as implemented here) probably only makes sense on pre-trained weights, to determine the best weight-sharing possible based on the trained parameters. Our experience so far shows that it would be impractical to train a model from scratch with such a severe restriction on weight values from the get-go - the training is likely to either take too long or never succeed.

Adding support for some basic scheduling, as you are suggesting above, should help address this issue and improve the usability of the clustering API.